portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -28,33 +28,89 @@
-    ));
-    sampleComments.add(new Comment(
-      "Rich Goulaz",
-      "This is not an example of a comment... or maybe it is? How to tell..."
-    ));
-    sampleComments.add(new Comment(
-      "Bob Guy",
-      "My name is Bob. I like this page."
-    ));
-  }
+  private final List<Comment> sessionComments = new ArrayList<Comment>();
+  private final Gson gson = new Gson();
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
     // Convert comments to JSON
     Gson gson = new Gson();
-    String commentJson = gson.toJson(sampleComments);
+    String commentJson = gson.toJson(sessionComments);
 
     // Send the JSON as the response
     response.setContentType("application/json;");
     response.getWriter().println(commentJson);
   }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get new comment parameters
+    final String COMMENT_AUTHOR_PARAM = "author";
+    final String COMMENT_BODY_PARAM = "comment-body";
+    String commentAuthor = getParameter(request, COMMENT_AUTHOR_PARAM, "");
+    String commentBody = getParameter(request, COMMENT_BODY_PARAM, "");
+
+    // Remove unsafe characters
+    commentAuthor = commentAuthor.replaceAll(UNSAFE_CHARACTERS_REGEX, "");
+    commentBody = commentBody.replaceAll(UNSAFE_CHARACTERS_REGEX, "");
+
+    // Validate the comment parameters
+    String validationError = validateIncomingComment(commentAuthor, commentBody);
+    if (validationError != null) {
+      response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+      response.setContentType("text/html;");
+      response.getWriter().println(validationError);
+      return;
+    }
+
+    // Add the comment to the session comments
+    Comment newComment = new Comment(commentAuthor, commentBody);
+    sessionComments.add(newComment);
+
+    String commentJson = gson.toJson(newComment);
+
+    // Return the new comment
+    response.setStatus(HttpServletResponse.SC_CREATED);
+    response.setContentType("application/json;");
+    response.getWriter().println(commentJson);
+    response.sendRedirect("/index.html");
+  }
+
+  /**
+   * Validates the comment parameters of a new comment.
+   * @return Return the error string or null if there are no errors.
+   */
+  private String validateIncomingComment(String commentAuthor, String commentBody) {
+    String validationErrors = "";
+
+    // Check if author or body fields are blank
+    if (commentAuthor.isBlank()) {
+      validationErrors += "Please include a comment author (cannot be whitespace).";
+    }
+    if (commentBody.isBlank()) {
+      if (validationErrors.length() > 0) validationErrors += " ";
+      validationErrors += "Please include a comment body (cannot be whitespace).";
+    }
+
+    if (validationErrors.isEmpty()) {
+      return null;
+    }
+    else {
+      return validationErrors;
+    }
```

Comment:
```suggestion
    if (validationErrors.isEmpty()) {
      return null;
    } else {
      return validationErrors;
    }
```

In general people write else statements on the same line as the closing `}` for the if statement. The javascript style guide doesn't mention this explicitly, but they do follow this style in the guide (like in the code here: https://google.github.io/styleguide/jsguide.html#formatting-function-expressions for example)

Reply:
Oh wow okay that one's new to me and this probably wont be the last time I make that mistake with muscle memory and all haha. Going to do a pass and make sure there aren't others.